feat: Minify shaders in runtime (#2801)
software-mansion/TypeGPU: apps/typegpu-docs/src/content/docs/advanced/minifying-shaders.mdx

Regular JS minification & obfuscation does not alter the shader code generated by TypeGPU.
This is why we provide specific options for transforming shaders.

## Mangling variable names in plugin

`unplugin-typegpu` collects function metadata, which is later used for WGSL code generation.
With the `{ EXPERIMENTAL_obfuscate: true }` option, all saved identifiers will be obfuscated:
- In the AST, all parameters and variables will have their names changed to `a`, `b`, `c`, ...
- Externals (the captured scope used by the function) will also have their respective identifiers changed.

:::note
It is required to disable the plugin auto-naming by setting `{ autoNamingEnabled: false }` for obfuscation to work. 

> Only resources given name via `.$name()` will be named in the resulting shader.

Reviewer: ```suggestion
The name given to a resource via `.$name()` will not be obfuscated in the resulting shader.
```
Technically speaking, other resources are still named but with pseudo-random letters.

- In the AST, all parameters and variables will have their names changed to `a`, `b`, `c`, ...
- Externals (the captured scope used by the function) will also have their respective identifiers changed.

:::note
It is required to disable the plugin auto-naming by setting `{ autoNamingEnabled: false }` for obfuscation to work. 
Only resources given name via `.$name()` will be named in the resulting shader.
:::

:::caution
Enabling obfuscation will also obfuscate error messages appearing during resolution.

It is not advised to minify shaders during development.
:::

As an example, take a look at the following function:
```ts twoslash
import { tgpu, d } from 'typegpu';
// ---cut---
const coneVolume = tgpu.fn([d.f32, d.f32], d.f32)((radius, height) => {
  'use gpu';
  const baseArea = Math.PI * radius ** 2;
  const volume = (baseArea * height) / 3;
  return volume;
});
```

When obfuscated and resolved, it will look like this:
```wgsl
fn item(a: f32, b: f32) -> f32 {
  let c = (3.141592653589793f * pow(a, 2f));
  let e = ((c * b) / 3f);
  return e;
}
```

:::note
Obfuscation does not affect raw-WGSL implemented functions.
:::

## Runtime minification

> Runtime minification removes all comments (possibly present in raw-WGSL implemented functions), and all redundant whitespaces.

Author: It will be merged soon enough
